### PR TITLE
Fix AssertSet can't verify the occurrence of a property set when a result of a mocked object is used as a value

### PR DESCRIPTION
**Pull request by ivo-stoilov**
o enabled dispatching invocation to mocks
o introduced new property in Invocation indicating AssertSet context
o adjusted some behavior processing in AssertSet scope
o added unit tests to cover the change

### CODE REVIEW

File: Telerik.JustMock/Core/MocksRepository.cs [hunk 1]
```diff
@@ -415,15 +416,18 @@ internal void DispatchInvocation(Invocation invocation)
...
 			if (!methodMockProcessed)
 			{
+				// We have to be cearfull for the potential exception throwing in the assertion context, so skip CallOriginalBehavior processing
 				var mock = invocation.MockMixin;
 				if (mock != null)
 				{
-					foreach (var behavior in mock.FallbackBehaviors)
-						behavior.Process(invocation);
+                    mock.FallbackBehaviors
+                        .Where(behavior => !invocation.InAssertSet || !(behavior is CallOriginalBehavior))
```
Comment (mihail-vladov) [inline Telerik.JustMock/Core/MocksRepository.cs]: It seems that there is no unit test for this scenario when the CallOriginalBehavior should not be processed.

File: Telerik.JustMock/Core/MocksRepository.cs [hunk 2]
```diff
@@ -415,15 +416,18 @@ internal void DispatchInvocation(Invocation invocation)
...
+                        .ToList()
+                        .ForEach(behavior => behavior.Process(invocation));
```
Comment (mihail-vladov) [inline Telerik.JustMock/Core/MocksRepository.cs]: As we talked a few days earlier, LINQ queries are hard to debug. To improve the debug experience a variable can be introduced with the filtered behaviors for processing.

File: Telerik.JustMock/Core/MocksRepository.cs [hunk 3]
```diff
@@ -803,11 +807,23 @@ internal void AssertAction(string message, Action memberAction, Args args = null
 		{
 			using (MockingContext.BeginFailureAggregation(message))
 			{
-				var callPattern = ConvertActionToCallPattern(memberAction);
+				var callPattern = ConvertActionToCallPattern(memberAction, true);
```
Comment (mihail-vladov) [inline Telerik.JustMock/Core/MocksRepository.cs]: The public asserts API using this method are not fully covered with unit tests and allowing dispatch to method mocks can be dangerous.

File: Telerik.JustMock/Core/MocksRepository.cs (hunk 1 above)
Comment (ivo-stoilov) [reply Telerik.JustMock/Core/MocksRepository.cs]: @mihail-vladov, it seems that we have such, please see [ShouldNotThrowDuringAssertForCallOriginalWhenNoArrangeSpecified](https://github.com/telerik/JustMockLite/blob/463187c12bdf1b743fad63a4c8ac9188a638a36a/Telerik.JustMock.Tests/AssertionFixture.cs#L407)

File: Telerik.JustMock/Core/MocksRepository.cs (hunk 2 above)
Comment (ivo-stoilov) [reply Telerik.JustMock/Core/MocksRepository.cs]: @mihail-vladov, fixed with [fddf417](https://github.com/telerik/JustMockLite/pull/39/commits/fddf417f417b749148a45ea863e9b37874674e50)

File: Telerik.JustMock/Core/MocksRepository.cs (hunk 3 above)
Comment (ivo-stoilov) [reply Telerik.JustMock/Core/MocksRepository.cs]: @mihail-vladov, fixed with [fddf417](https://github.com/telerik/JustMockLite/pull/39/commits/fddf417f417b749148a45ea863e9b37874674e50)

**Pull request merged**